feat: Allow Return Invoices in Bank Reco

marination opened the pull request:
Fixes https://github.com/alyf-de/banking/issues/56

### Working
- Allow return invoices in the tool:
 	- Deposit: Allow SINV, Return SINV (discount) and Return PINV (refund)
 	- Withdrawal: Allow PINV, Return PINV and Return SINV
- Allow using discount (return) vouchers while matching:
	- Deposit: User can select a combination of -> [SINV, Return SINV] OR [just Return PINV].
	- Withdrawal: User can select a combination of -> [PINV, Return PINV] OR [just Return SINV].
- Show max limit of allocated amount and overflow in match tab summary (needs testing)
   <img width="800" alt="Screenshot 2024-09-12 at 6 57 52 PM" src="https://github.com/user-attachments/assets/7f776a0f-e1f4-4832-a7a1-c843a0ce377e">

marination committed:
feat: Allow Return Invoices in Bank Reco

- WIP

marination committed:
fix: Handle negative and positive invoice amounts (wip)

- TODO: tests

marination committed:
Merge branch 'version-14' into return-inv

marination committed:
fix: Missing argument

barredterra committed:
Merge branch 'version-14' into return-inv

barredterra commented on banking/overrides/bank_transaction.py:
```diff
@@ -176,6 +195,47 @@ def prepare_invoices_to_split(self, invoices):
+			# If there is only one type of amount, then dont bother adjusting
+			return sum_positive, sum_negative
+
+		difference = flt(sum_positive - sum_negative, self.precision("unallocated_amount"))
+		if (sum_positive < self.unallocated_amount) or (difference > self.unallocated_amount):
+			# NOTE: If sum_positive is far smaller than unallocated, avoid further reduction ((100, -10), 130)
+			# If difference is greater than unallocated (sum_positive is far greater), avoid further reduction ((1000, -10), 130)
+			# In both cases we want to reach closer to unallocated_amount
+			# Hence, sum_negative reduction is not pointless
+			sum_negative = 0
+		elif difference < self.unallocated_amount:
+			# Reduce the negative sum
+			to_reduce = difference
+			if difference >= sum_negative:
+				to_reduce = flt(
+					self.unallocated_amount - difference, self.precision("unallocated_amount")
+				)
+			sum_negative = flt(sum_negative - to_reduce, self.precision("unallocated_amount"))
```
This part leads to a "Debit/Credit Mismatch" error in my test case. If I just comment it out, it works perfectly.

I tested with 3 Sales Invoice and 4 Credit Notes, the balance of which was paid with a single bank transaction.

marination committed:
fix: Simplify +ve and -ve amount allocation

- Allocation of purely -ve amounts is not allowed. The goal is to knock of the bank transaction. We cannot end up owing money in this flow. That could happen elsewhere
- Allocation of +ve amounts follows the same olf flow of allocating just as much is needed
- In a moxed case, make sure +ve > -ve (point 1) and if this total is overallocating overall, reduce this excess from the positive amount

marination committed:
refactor: Invoice queries abstractions (generalise kwargs)

marination committed:
fix: Allow -ve only invoices to be reconciled for refund reasons

marination committed:
fix: Show max allocated

barredterra committed:
test: invoices and returns paid by one bank transaction

barredterra commented:
@Mergifyio backport version-15
